### ISSUE TITLE
symfony/security - tagging UserCheckerInterface

### ISSUE THREAD

**Issue by iltar** [symfony/symfony#11090]
Security - tagging UserCheckerInterface
========================================

When working with custom authentications, you most likely want to implement a custom UserChecker. The standard "form" supports this but doesn't really allow customization. "simple_form" does not have a standard implementation for this, sadly.

Proposal
========

I came to the conclusion that it might be a lot easier to add some custom account/user validation. I was creating my own Authentication and user provider and stumbled against the choice to implement the Simple variant or the normal (older) variant. 

One of the main reasons I couldn't implement the existing variant, was due to the lack of customization of the UserChecker. Before loading the user, I had to know if the user was supposed to see a captcha or if the IP of the user was blocked. Before the authentication, I wanted to know if the account had any blocks. After the authentication, I wanted to load a secondary Entity that I would check some additional account statuses. I could not work around this with existing code so had to write a custom provider for this.

A simple solution would be to allow the firewall options for a certain authentication method to allow passing the ID of a custom service. This would mean you'd get the following option:

```yml
security:
    firewall:
        secured_area:
            (simple_)form:
                user_checker_service: my.user_checker
```

As long as your service would implement the UserChecker Interface, this would all be possible. However, this isn't really flexible... You could do the following:

```yml
security:
    firewall:
        secured_area:
            (simple_)form:
                user_checkers: 
                    - my.user_checker.ip_block
                    - my.user_checker.captcha_block
                    - my.user_checker.banned_account
                    - my.user_checker.accepts_latest_eula
                    - vendor.bundle.extra_user_check
```

With this solution, you would be able to add extra checks on demand. It would be a really nice addition to work with the Simple provider as well, but that would introduce BC breaks, but solve a lot of nasty code.

```php
class MyBannedAccountUserChecker implements
    PreUserLoadCheckerInterface,
    PreAuthUserCheckerInterface, 
    PostAuthUserCheckerInterface
{
    public function checkPreUserLoad()
    {
        // additional checks BEFORE the user is being load...
        // good example is to check IP block, captcha's, other requirements
    }

    public function checkPreAuth(UserInterface $user)
    {
        if (!$user instanceof MyUser) {
            return;
        }
        
        // .. get the entity where ever you store if the user is banned
        
        if ($banned) {
            throw new AccountBannedException($user); // extends AuthenticationException
        }
    }
    
    public function checkPostAuth(UserInterface $user) { ... }
}
```
I think this example should demonstrate exactly what I mean. This will give you flexibility while it will also make it easier to configure the security component user checks. I have noticed examples of people who are totally lost in this component and have no clue where to add the custom checks for this. As far as I know it's only possible to modify the current existing UserChecker by a CompilerPass. It's tightly coupled with the "form" login method and the only available implementation works only with the AdvancedUserInterface (which imo should be rm -rf'd).

To prevent backwards incompatibility issues, the current UserCheckerInterface should not be removed, although the splitting of pre and post auth checks can cause a minor inconvenience.

**Comment by lscholten** [comment]
I would also like a possibility to attach UserCheckers to my firewalls. I've tried overriding the standard UserChecker, but I found that dirty. And, more importantly, as iltar said, "simple_form" does not call any UserChecker (which also has the consequence that AdvancedUserInterface does nothing when working with "simple_form").

My use case is that I need to be able to block the User from logging in based on some of their attributes. However, the check is dependent on a service in the container, so I would like to detach this logic from the User class. I found the UserChecker to be the perfect concept for this kind of problem.

**Comment by weaverryan** [comment]
@iltar I like this idea and certainly see the problem. Some notes (I have sone -1's, but you'll see it's just because I'm trying to focus us!)

1) Security is complex, so let's keep any proposed changes as small as possible :)
2) I'm :-1: on the events - it's a bigger change and once again would allow 3rd-party bundles to automatically register listeners
3) I'm :-1: on splitting the interface - I think it will cause some complications without too much benefit (again, the smaller the changes, the better chance of getting a great solution through!). Anyways, the interface is 2 methods, so if you don't need to do something in one method, just leave it blank.
4) I'm :-1: also on `checkPreUserLoad` for the same reason as above. But I *do* understand there's a legit use-case here. I think we should solve it another way - like a normal event listener (not sure if that's possible) or some other key under the firewall that allows you to have a listener really early on in the firewall process. The `UserChecker` is for checking users - this feels more like a global "firewall" type of thing to me.
5) The `security.user_checker` is global in the system currently. If we were to make it configurable, I think it should be done at the firewall level, not all the way at the authentication provider level. Sure, we could go further later if there's a real use-case.
6) I think there should only be one use checker - not an array. If you want multiple, you can register one, then internally call out to other services (or dispatch an event, or do whatever you want).

In conclusion: Keep the interface the same, but add a new setting under the firewall that allows you to configure the "user checker". For BC, we can keep the `security.user_checker` service and only change the user checker that's injected if it's configured under the firewall.

Thanks!

**Comment by iltar** [comment]
@weaverryan I agree with the complexity and adding stuff. I can live with the fact that `preUserLoad` might not be the best. The only thing I'm afraid of, is that I will constantly have to update my user checker for every pass (which might become relatively large) and this becomes a mess to write unit-tests for. It also stands in the way of code re-usability. If I want to re-use my IP check for example, I will have either duplicate a lot of code, or use a custom composition over inheritance method.

I'm afraid that with the composition part, it will still be very hard to add generic solutions to the security, such as an IP block on X attempt. Ideally I'd create a bundle that provides configurable behavior and all a developer has to do, is add the checker to the firewall after enabling the bundle.

**Comment by weaverryan** [comment]
@iltar Hmm, fair point about being able to reuse little "user checker" implementations. We can try the "array of user checkers" then and see what others think. Implementing a "chain" for this is easy - just trying to keep core as simple as possible.

Now, @iltar, are you brave enough to want to work on this? ;)

**Comment by iltar** [comment]
@weaverryan After my work on the `SecurityContext`, I consider this a holiday job :+1:

I'll provide a `ChainUserChecker` which uses the `UserCheckerInterface` so it integrates seamlessly and is fully BC. I'm just not sure what the desirable security config would be. I could put it optionally under a firewall, but I'd have to support all current (non-simple) implementation that currently use the `SecurityChecker`.

Personally I was thinking of this as default values for the BC situation, this should make it compatible for 2.3 up to 2.8 configs. We can then later on decide if 3.0 should follow the same behavior. This array can then by inserted in the `ChainUserProvider` and walk through it in the same sequence.

```yml
security:
    firewall:
        secured_area:
            form_login:
                user_checkers: 
                    - security.user_checker
```

_In this case I've dropped the service tags, but if desired, that can always be added in the future._

The current implementation only works if you extend the `User` class or use the `AdvancedUserInterface`. I think this should all be revised for 3.0, but IMO this should go (the method names are terribly confusing).

**Comment by iltar** [comment]
@weaverryan I've done the first commit and will try to work on it frequently. In case you're interested: https://github.com/iltar/symfony/commit/0058d60d5bab020d2801e44784a2cc4ffd675ed0

**Comment by weaverryan** [comment]
@iltar For the configuration, I would do whatever feels most natural at first, then we can tweak it later. What I mean is, this stuff is hard enough - getting *something* working as soon as possible is a good goal. Then we can discuss the details. But, I will say that my vote is still to support it at the firewall level (so not under `form_login` or other things, or maybe *also* there, in addition to under the firewall). That would be consistent with the "user provider" (which is actually under both). This might be trickier, however, because - as you said - you'll need to make sure it works with all existing auth providers that use the "user checker". 

Ping me as progress goes - I saw the first commit :).

**Comment by iltar** [comment]
@weaverryan It wasn't too hard to add the configuration underneath the firewall, especially as the existing tests are easy to expand on. I'm not entirely sure about the xml config, if this is as nice due to the plural config, didn't manage to fix that. https://github.com/iltar/symfony/commit/3f016d7842462339df4e1a69ceb91b01526aedc9

All that's left to do, is hook up the `security.change_user_checker.$id` to their respectful authenticators and then it should be done, more or less.

**Comment by iltar** [comment]
@weaverryan As it was easier than I thought: https://github.com/iltar/symfony/commit/1408743b36ef6e481c628a86c327b48dce706cb6. This commit adds the chain user checker.

Did I miss anything for this feature?
